In [7]:
import pandas as pd

df = pd.read_csv(r'C:\Users\clint\Desktop\RA Task\extracted_bidders_all.csv')
df

project_number                        bidder_name  \
0            180003                  RONYAK PAVING INC   
1            180003                 SHELLY & SANDS INC   
2            180003              KOSKI CONSTRUCTION CO   
3            180003  KOKOSING CONSTRUCTION COMPANY INC   
4            180003          CHAGRIN VALLEY PAVING INC   
..              ...                                ...   
592          188000               M & B ASPHALT CO INC   
593          188000                BLUFFTON PAVING INC   
594          188000                     SHELLY COMPANY   
595          188001                     SHELLY COMPANY   
596          188001                 SHELLY & SANDS INC   

                     address           city state         zip    county  \
0        14376 N CHESHIRE ST         BURTON    OH       44021    Geauga   
1            1515 HARMON AVE       COLUMBUS    OH       43223  Franklin   
2               P O BOX 1038      ASHTABULA    OH  44005-1038       NaN   
3          886 MC KINLEY AVE       COLUMBUS    OH       43222  Franklin   
4              17290 MUNN RD  CHAGRIN FALLS    OH       44023    Geauga   
..                       ...            ...   ...         ...       ...   
592  1525 WEST COUNTY RD #42         TIFFIN    OH       44883    Seneca   
593               P O BOX 26       BLUFFTON    OH       45817     Allen   
594       80 PARK DR BOX 266     THORNVILLE    OH       43076     Perry   
595       80 PARK DR BOX 266     THORNVILLE    OH       43076     Perry   
596          1515 HARMON AVE       COLUMBUS    OH       43223  Franklin   

     bidder_number  bid_amount          filename  awarded  
0                1  2087863.70  180003bidtab.pdf     True  
1                2  2193929.12  180003bidtab.pdf    False  
2                3  2268022.35  180003bidtab.pdf    False  
3                4  2298511.32  180003bidtab.pdf    False  
4                5  2400000.00  180003bidtab.pdf    False  
..             ...         ...               ...      ...  
592              1   918732.90  188000bidtab.pdf     True  
593              3   984896.65  188000bidtab.pdf    False  
594              4   988535.87  188000bidtab.pdf    False  
595              1  2049892.28  188001bidtab.pdf     True  
596              2  2102212.85  188001bidtab.pdf    False  

[597 rows x 11 columns]

In [8]:
df2 = pd.read_csv(r'C:\Users\clint\Desktop\RA Task\7.csv')
df2

state    county   fips  year        project_start  project_id  \
0    Ohio  Paulding  39125  2018  2018-05-24 00:00:00      105522   
1    Ohio   Wyandot  39175  2018  2018-11-15 00:00:00       88832   
2    Ohio    Butler  39017  2018  2018-01-18 00:00:00       94263   
3    Ohio  Franklin  39049  2018  2018-01-18 00:00:00       76467   
4    Ohio   Hocking  39073  2018  2018-01-18 00:00:00      101555   
..    ...       ...    ...   ...                  ...         ...   
198  Ohio  Harrison  39067  2018  2018-12-20 00:00:00       91844   
199  Ohio  Highland  39071  2018  2018-12-20 00:00:00       84622   
200  Ohio   Madison  39097  2018  2018-12-20 00:00:00      105547   
201  Ohio   Clinton  39027  2018  2018-12-20 00:00:00       87300   
202  Ohio   Belmont  39013  2018  2018-12-20 00:00:00       91962   

              route  mileage  lanes  project_duration_days  eng_estimate_mils  \
0               111   12.982    2.0                   99.0              0.943   
1                23      NaN    4.0                  290.0              3.702   
2                73      NaN    2.0                  195.0              0.287   
3         270 / 315      NaN    4.0                  255.0              4.904   
4                33      NaN    2.0                  255.0              0.435   
..              ...      ...    ...                    ...                ...   
198         250 / 9      NaN    2.0                  284.0              2.249   
199       138 / 753      NaN    2.0                  223.0              1.508   
200              42      NaN    2.0                  284.0              3.727   
201  251 / 68 / 350      NaN    2.0                  589.0              4.893   
202             800      NaN    2.0                  345.0              5.605   

     win_bid_mils  cost_mils  num_bidders                   bidders_list  \
0        0.957859   1.047510          2.0  Shelly Company, Gerken Paving   
1        3.236775   3.304783          3.0                            NaN   
2        0.258900   0.232678          3.0                            NaN   
3        6.101481   6.991613          3.0                            NaN   
4        0.553756   0.531749          2.0                            NaN   
..            ...        ...          ...                            ...   
198      2.284000   2.212447          3.0                            NaN   
199      1.494436   1.658765          2.0                            NaN   
200      3.611668   3.535001          4.0                            NaN   
201      5.441520   5.331566          1.0                            NaN   
202      5.165815   5.117317          2.0                            NaN   

                                            all_routes  cost_overrun_pct  \
0                                                  NaN          9.359511   
1                                   ['23', '23', '23']          2.101101   
2                                               ['73']        -10.128362   
3    ['270', '270', '270', '270', '270', '270', '27...         14.588788   
4                                               ['33']         -3.974212   
..                                                 ...               ...   
198                                       ['9', '250']         -3.132812   
199                       ['138', '138', '138', '753']         10.996011   
200                                 ['42', '42', '42']         -2.122762   
201  ['68', '68', '68', '68', '68', '68', '251', '3...         -2.020640   
202                              ['800', '800', '800']         -0.938836   

     Project Num  
0         180326  
1         180569  
2         180006  
3         180012  
4         180020  
..           ...  
198       180609  
199       180610  
200       180611  
201       180621  
202       180622  

[203 rows x 18 columns]

In [9]:
import re

def clean_bidder_name(name):
    # Convert to title case
    name = name.title()
    # Remove common company suffixes for professionalism
    name = re.sub(r'\b(Inc|Llc|Corp|Corporation|Limited|Ltd|Co|Company|Incorporated)\b[., ]*', '', name, flags=re.IGNORECASE)
    # Remove extra spaces and commas
    name = re.sub(r'\s{2,}', ' ', name).strip()
    name = re.sub(r',\s*,', ',', name)
    name = name.strip(', ')
    return name

# Ensure 'bidders_list' column exists in df2
if 'bidders_list' not in df2.columns:
    df2['bidders_list'] = ''

for idx, row in df.iterrows():
    project_number = row['project_number']
    bidder_name = clean_bidder_name(str(row['bidder_name']))
    # Find matching rows in df2
    match = df2['Project Num'] == project_number
    for i in df2[match].index:
        current = df2.at[i, 'bidders_list']
        if pd.isna(current) or current == '':
            df2.at[i, 'bidders_list'] = bidder_name
        else:
            # Split, clean, and rejoin to avoid duplicates and ensure formatting
            names = [clean_bidder_name(n) for n in current.split(',') if n.strip()]
            if bidder_name not in names:
                names.append(bidder_name)
            df2.at[i, 'bidders_list'] = ', '.join(names)

df2

state    county   fips  year        project_start  project_id  \
0    Ohio  Paulding  39125  2018  2018-05-24 00:00:00      105522   
1    Ohio   Wyandot  39175  2018  2018-11-15 00:00:00       88832   
2    Ohio    Butler  39017  2018  2018-01-18 00:00:00       94263   
3    Ohio  Franklin  39049  2018  2018-01-18 00:00:00       76467   
4    Ohio   Hocking  39073  2018  2018-01-18 00:00:00      101555   
..    ...       ...    ...   ...                  ...         ...   
198  Ohio  Harrison  39067  2018  2018-12-20 00:00:00       91844   
199  Ohio  Highland  39071  2018  2018-12-20 00:00:00       84622   
200  Ohio   Madison  39097  2018  2018-12-20 00:00:00      105547   
201  Ohio   Clinton  39027  2018  2018-12-20 00:00:00       87300   
202  Ohio   Belmont  39013  2018  2018-12-20 00:00:00       91962   

              route  mileage  lanes  project_duration_days  eng_estimate_mils  \
0               111   12.982    2.0                   99.0              0.943   
1                23      NaN    4.0                  290.0              3.702   
2                73      NaN    2.0                  195.0              0.287   
3         270 / 315      NaN    4.0                  255.0              4.904   
4                33      NaN    2.0                  255.0              0.435   
..              ...      ...    ...                    ...                ...   
198         250 / 9      NaN    2.0                  284.0              2.249   
199       138 / 753      NaN    2.0                  223.0              1.508   
200              42      NaN    2.0                  284.0              3.727   
201  251 / 68 / 350      NaN    2.0                  589.0              4.893   
202             800      NaN    2.0                  345.0              5.605   

     win_bid_mils  cost_mils  num_bidders  \
0        0.957859   1.047510          2.0   
1        3.236775   3.304783          3.0   
2        0.258900   0.232678          3.0   
3        6.101481   6.991613          3.0   
4        0.553756   0.531749          2.0   
..            ...        ...          ...   
198      2.284000   2.212447          3.0   
199      1.494436   1.658765          2.0   
200      3.611668   3.535001          4.0   
201      5.441520   5.331566          1.0   
202      5.165815   5.117317          2.0   

                                          bidders_list  \
0                                Shelly, Gerken Paving   
1        Shelly, Shelly & Sands, Kokosing Construction   
2    Barrett Paving Materials, John R Jurgensen, Ra...   
3        Kokosing Construction, Shelly, Shelly & Sands   
4                               Shelly & Sands, Shelly   
..                                                 ...   
198  Nls Paving, Shelly & Sands, Barbicas Construction   
199                           Shelly, John R Jurgensen   
200  Shelly & Sands, Kokosing Construction, Shelly,...   
201                                   John R Jurgensen   
202                         Shelly & Sands, Nls Paving   

                                            all_routes  cost_overrun_pct  \
0                                                  NaN          9.359511   
1                                   ['23', '23', '23']          2.101101   
2                                               ['73']        -10.128362   
3    ['270', '270', '270', '270', '270', '270', '27...         14.588788   
4                                               ['33']         -3.974212   
..                                                 ...               ...   
198                                       ['9', '250']         -3.132812   
199                       ['138', '138', '138', '753']         10.996011   
200                                 ['42', '42', '42']         -2.122762   
201  ['68', '68', '68', '68', '68', '68', '251', '3...         -2.020640   
202                              ['800', '800', '800']         -0.938836   

     Project Num  
0         180326  
1         180

In [10]:
# Check robustness: Compare number of bidders in 'bidders_list' to 'num_bidders' in df2

def count_bidders(bidders):
    if pd.isna(bidders) or bidders.strip() == '':
        return 0
    return len([b.strip() for b in bidders.split(',') if b.strip()])

# Calculate the number of bidders in 'bidders_list'
df2['bidders_count'] = df2['bidders_list'].apply(count_bidders)

# Find rows where the count does not match 'num_bidders'
mismatches = df2[df2['bidders_count'] != df2['num_bidders']]

if mismatches.empty:
    print('All entries are consistent: number of bidders matches num_bidders.')
else:
    print('Rows with mismatched bidder counts:')
    display(mismatches[['Project Num', 'bidders_list', 'bidders_count', 'num_bidders']])

Rows with mismatched bidder counts:


Project Num                            bidders_list  bidders_count  \
8       188000  M & B Asphalt, Bluffton Paving, Shelly              3   

   num_bidders  
8          4.0

We see there is one error only out of 202 projects. We will manuall correct this. 

In [11]:
# Manually add 'HELMS AND SONS EXCAVATING' to bidders_list for Project Num 188000 in df2
project_num = 188000
bidder_to_add = 'HELMS AND SONS EXCAVATING'

def clean_bidder_name(name):
    import re
    name = name.title()
    name = re.sub(r'\b(Inc|Llc|Corp|Corporation|Limited|Ltd|Co|Company|Incorporated)\b[., ]*', '', name, flags=re.IGNORECASE)
    name = re.sub(r'\s{2,}', ' ', name).strip()
    name = re.sub(r',\s*,', ',', name)
    name = name.strip(', ')
    return name

# Find the row(s) in df2 with Project Num 188000
mask = df2['Project Num'] == project_num
for i in df2[mask].index:
    current = df2.at[i, 'bidders_list']
    new_bidder = clean_bidder_name(bidder_to_add)
    if pd.isna(current) or current == '':
        df2.at[i, 'bidders_list'] = new_bidder
    else:
        names = [clean_bidder_name(n) for n in current.split(',') if n.strip()]
        if new_bidder not in names:
            names.append(new_bidder)
        df2.at[i, 'bidders_list'] = ', '.join(names)

# Display the updated row(s) for verification
df2.loc[mask, ['Project Num', 'bidders_list']]

Project Num                                       bidders_list
8       188000  M & B Asphalt, Bluffton Paving, Shelly, Helms ...

In [12]:
# Check robustness: Compare number of bidders in 'bidders_list' to 'num_bidders' in df2

def count_bidders(bidders):
    if pd.isna(bidders) or bidders.strip() == '':
        return 0
    return len([b.strip() for b in bidders.split(',') if b.strip()])

# Calculate the number of bidders in 'bidders_list'
df2['bidders_count'] = df2['bidders_list'].apply(count_bidders)

# Find rows where the count does not match 'num_bidders'
mismatches = df2[df2['bidders_count'] != df2['num_bidders']]

if mismatches.empty:
    print('All entries are consistent: number of bidders matches num_bidders.')
else:
    print('Rows with mismatched bidder counts:')
    display(mismatches[['Project Num', 'bidders_list', 'bidders_count', 'num_bidders']])

All entries are consistent: number of bidders matches num_bidders.


We see our data is consistent now. 

In [13]:
df2

state    county   fips  year        project_start  project_id  \
0    Ohio  Paulding  39125  2018  2018-05-24 00:00:00      105522   
1    Ohio   Wyandot  39175  2018  2018-11-15 00:00:00       88832   
2    Ohio    Butler  39017  2018  2018-01-18 00:00:00       94263   
3    Ohio  Franklin  39049  2018  2018-01-18 00:00:00       76467   
4    Ohio   Hocking  39073  2018  2018-01-18 00:00:00      101555   
..    ...       ...    ...   ...                  ...         ...   
198  Ohio  Harrison  39067  2018  2018-12-20 00:00:00       91844   
199  Ohio  Highland  39071  2018  2018-12-20 00:00:00       84622   
200  Ohio   Madison  39097  2018  2018-12-20 00:00:00      105547   
201  Ohio   Clinton  39027  2018  2018-12-20 00:00:00       87300   
202  Ohio   Belmont  39013  2018  2018-12-20 00:00:00       91962   

              route  mileage  lanes  project_duration_days  eng_estimate_mils  \
0               111   12.982    2.0                   99.0              0.943   
1                23      NaN    4.0                  290.0              3.702   
2                73      NaN    2.0                  195.0              0.287   
3         270 / 315      NaN    4.0                  255.0              4.904   
4                33      NaN    2.0                  255.0              0.435   
..              ...      ...    ...                    ...                ...   
198         250 / 9      NaN    2.0                  284.0              2.249   
199       138 / 753      NaN    2.0                  223.0              1.508   
200              42      NaN    2.0                  284.0              3.727   
201  251 / 68 / 350      NaN    2.0                  589.0              4.893   
202             800      NaN    2.0                  345.0              5.605   

     win_bid_mils  cost_mils  num_bidders  \
0        0.957859   1.047510          2.0   
1        3.236775   3.304783          3.0   
2        0.258900   0.232678          3.0   
3        6.101481   6.991613          3.0   
4        0.553756   0.531749          2.0   
..            ...        ...          ...   
198      2.284000   2.212447          3.0   
199      1.494436   1.658765          2.0   
200      3.611668   3.535001          4.0   
201      5.441520   5.331566          1.0   
202      5.165815   5.117317          2.0   

                                          bidders_list  \
0                                Shelly, Gerken Paving   
1        Shelly, Shelly & Sands, Kokosing Construction   
2    Barrett Paving Materials, John R Jurgensen, Ra...   
3        Kokosing Construction, Shelly, Shelly & Sands   
4                               Shelly & Sands, Shelly   
..                                                 ...   
198  Nls Paving, Shelly & Sands, Barbicas Construction   
199                           Shelly, John R Jurgensen   
200  Shelly & Sands, Kokosing Construction, Shelly,...   
201                                   John R Jurgensen   
202                         Shelly & Sands, Nls Paving   

                                            all_routes  cost_overrun_pct  \
0                                                  NaN          9.359511   
1                                   ['23', '23', '23']          2.101101   
2                                               ['73']        -10.128362   
3    ['270', '270', '270', '270', '270', '270', '27...         14.588788   
4                                               ['33']         -3.974212   
..                                                 ...               ...   
198                                       ['9', '250']         -3.132812   
199                       ['138', '138', '138', '753']         10.996011   
200                                 ['42', '42', '42']         -2.122762   
201  ['68', '68', '68', '68', '68', '68', '251', '3...         -2.020640   
202                              ['800', '800', '800']         -0.938836   

     Project Num  bidders_count  
0         180326 

In [14]:
# Save df2 as 10.csv in the current directory
df2.to_csv('10.csv', index=False)